ma_gym/envs/traffic_junction/traffic_junction.py

```diff
@@ -125,25 +271,171 @@ def __create_grid(self):
+        """
+        Performs an action in the environment and steps forward. At each step a new agent enters the road by
+        one of the 4 gates according to a probability "_arrive_prob". A "ncoll" reward is given to an agent if it
+        collides and all of them receive "-0.01*step_n" to avoid traffic jams.
+
+        :param agents_action: list of actions of all the agents to perform in the environment
+        :type agents_action: list
+
+        :return: agents observations, rewards, if agents are done and additional info
+        :rtype: tuple
+        """
+        assert len(agents_action) == self.n_agents, \
+            "Invalid action! It was expected to be list of {}" \
+            " dimension but was found to be of {}".format(self.n_agents, len(agents_action))
+
+        # Todo: Add check to validate each sub-action. It should be a valid action.
+        # This must be done before position of any car is updated
+
+        self._step_count += 1
+        rewards = [0 for _ in range(self.n_agents)]  # initialize rewards array
+
+        # checks if there is a collision; this is done in the __update_agent_pos method
+        for agent_i, action in enumerate(agents_action):
+            if not (self._agent_dones[agent_i]) and self._on_the_road[agent_i]:
+                col_reward = self.__update_agent_pos(agent_i, action)
+                rewards[agent_i] += col_reward
+
+        # TODO remove self.curr_cars_count < self._n_max here, might not be needed if we play with the on the road
+        # adds new car according to the probability _arrive_prob
+        if random.uniform(0, 1) < self._arrive_prob and self.curr_cars_count < self._n_max:
+            for agent_i in range(self.n_agents):
+                if self.__is_gate_free() and not self._on_the_road[agent_i]:
+                    while True:
+                        pos = random.choice(list(self._route_vectors.keys()))
+                        self._agents_direction[agent_i] = self._route_vectors[pos]
+                        if self._is_cell_vacant(pos):
+                            self.agent_pos[agent_i] = pos
+                            self.curr_cars_count += 1
+                            self._on_the_road[agent_i] = True
+                            self._agent_turned[agent_i] = False
+                            self._agents_routes[agent_i] = random.randint(1, 3)
+                            self.__update_agent_view(agent_i)
+                            break
+
+        # verifies if any destination place was reached
+        for agent_i in range(self.n_agents):
+            if self.__reached_dest(agent_i):
+                self._on_the_road[agent_i] = False
+                self.curr_cars_count -= 1
+
+        if self._step_count >= self._max_steps:
+            for i in range(self.n_agents):
+                self._agent_dones[i] = True
+
+        # gives additional step punishment to avoid jams
+        for agent_i in range(self.n_agents):
```

Comment:
reward information is spread across 2 methods i.e over here(line 330) and  inside __update_agent_pos ( collision_reward).

Let's keep it in one place only. 

`Option 1`: move the step_cost inside update_agent_pos. 
`Option 2`: return collision as a "bool" from __update_agent_pos and then based on collision flag add the collision_reward.